### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by sarbbottam**
The binary interface has not been changed. However if one would prefer to `require` the module, below exposed APIs are available.
- `getCurrentRules // get all the current rules instead of referring the extended files or documentation`
- `getPluginRules // get all the plugin rules instead of referring the extended files or documentation`
- `getAllAvailableRules // get all the availale rules instead of referring eslint and pluging packages or documentation`
- `getNewRules`

I look forward to your concerns. 🐰 Happy Easter! 🐰

### CODE REVIEW

**Comment by ta2edchimp** [comment]
Personally, I'm all for exposing a module's method to be used as a lib :+1: 

But to be honest, I would have rather just `git mv ./bin-utils.js ./lib/rule-finder.js`, included `./index.js`'s content into `./lib/rule-finder.js` and adjusted package.json's `main` field.
I like the simple functional approach the methods utilize as of now, as well as that we expose the compiled `config` via `getConfig()` method.

So, I'm really unsure whether to refactor it exactly **this** way, and I tend towards declining it.
_May @kentcdodds decide ..._

**Comment by sarbbottam** [comment]
@ta2edchimp thanks for sharing your concerns.

> > I like the simple functional approach the methods

Me too, the proposed code is _pseudo-functional/pseudo-object-orieneted_, if I may say so. [All the computation is **purely functional**. It does not depends on the `state`. ](https://github.com/sarbbottam/eslint-find-new-rules/blob/refactor/src/rule-finder.js#L65-L70)

```
var configFile = _getConfigFile(specifiedFile)
var config = _getConfig(configFile)
var currentRules = _getCurrentRules(config)
var pluginRules = _getPluginRules(config)
var allRules = _getAllAvailableRules(pluginRules)
var newRules = difference(allRules, currentRules)
```

In fact, its similar to `bin-utils` combined with `index` along with following lines from `bin`

``` js
var config = binUtils.getConfig(process.argv[2])
var currentRules = binUtils.getCurrentRules(config)
var pluginRules = binUtils.getPluginRules(config)
```

The only reason to wrap the above stated computation in `constructor` is encapsulation.

The only drawback I see in the proposed approach is `two` calls.

``` js
var ruleFinder = new RuleFinder(specifiedFile)
var newRules = ruleFinder.getNewRules()
```

If any changes takes place between those two statement, for example, `eslint config` is updated after `var ruleFinder = new RuleFinder(specifiedFile)`, it would not take the changes in the account and `ruleFinder = new RuleFinder(specifiedFile)` needs to be _re-executed_.

**But it would be a rare scenario.**
Even in the current approach, one need to invoke more than one function and same could happen.

``` js
var config = binUtils.getConfig(process.argv[2])
var currentRules = binUtils.getCurrentRules(config)
var pluginRules = binUtils.getPluginRules(config)
```

**However we could stick to pure functional and always `accept` the `config file` as the argument.** But I guess the APIs need to be modified then.

**Comment by ta2edchimp** [comment]
Well I actually think of this approach to not take changes into account as a **real advantage**!
I could imagine scenarios, where you could use this to create diffs / track changes between different versions of a config file or sth. like that, when using this module as a lib.

Well, could we at least think of letting the module expose a factory function to return the desired `RuleFinder` object?
<sub>It's simply that I **personally** ... to keep it **very** brief ... don't like any appearances of the keyword `new`.</sub>

tl;dr, sth like this (shortened heavily):

``` JavaScript
function getRuleFinder(specifiedFile) {
  // the private stuff ...

  return {
    // prop shorthand notation only for demo purposes
    getCurrentRules() {
      return currentRules
    },
    getPluginRules() {
      return pluginRules
    },
    getAllAvailableRules() {
      return allRules
    },
    getNewRules() {
      return newRules
    }
  }
}

module.exports = getRuleFinder
```

I also think about whether something like `getUnusedRules` or `findUnconfiguredRules` would be a better and more concise name instead of `getNewRules` (as, you know, they don't have to necessarily be _new_, but are rather unconfigured / unused).

What do you think?

PS: Sorry if I may sound a bit ranting at times, it is absolutely not intended, but as not being a native speaker, it's just hard sometimes to express exactly what I mean :stuck_out_tongue_winking_eye:

**Comment by sarbbottam** [comment]
> > It's simply that I **personally** ... to keep it **very** brief ... don't like any appearances of the keyword **new**.

[**Me neither**, for `ES2015 class`, we have to use `new` and thus got back to the convention](https://github.com/airbnb/javascript/pull/246)

> > I also think about whether something like `getUnusedRules` or `findUnconfiguredRules` would be a better and more concise name instead of `getNewRules` (as, you know, they don't have to necessarily be new, but are rather `unconfigured` / `unused`).

:+1: Totally, I am bad with, naming things.

> > Well, could we at least think of letting the module expose a factory function to return the desired RuleFinder object?
> > 
> > ``` js
> > function getRuleFinder(specifiedFile) {
> >   ...
> >   return {
> >     ...
> >   }
> > }
> > module.exports = getRuleFinder
> > ```

How about?

``` js
module.exports = function(specifiedFile) {
  return new RuleFinder(specifiedFile)
}
```

**Comment by ta2edchimp** [comment]
I'd be fine with that!

Just for the sake of completeness, I'd do

``` JavaScript
module.exports = function getRuleFinder(specifiedFile) {
  return new RuleFinder(specifiedFile)
}
```

or `... = function ruleFinder(...)` or the like . . .

**Comment by sarbbottam** [comment]
@ta2edchimp could you take another look?

**Comment by ta2edchimp** [comment]
Will have a thorough look tomorrow (only on the phone atm), but looks good to me as far as I can tell now, and will probably going to merge, when @kentcdodds has nothing to add :smiley:

**Comment by ta2edchimp** [comment]
Some things that we have to consider prior to merging:
1. This introduces a **Breaking Change** (exposed API and its usage have changed).
   - Relevant commit messages should therefore be adjusted accordingly.
2. Documentation should reflect this change (that is, usage as a lib as a whole).

Had to track this here to not forget until tomorrow as I'm almost 😴

**Comment by sarbbottam** [comment]
- [ ] Update commit messages to reflect breaking changes
- [ ] Update documentation to reflect breaking changes/ usage

**Comment by sarbbottam** [comment]
@ta2edchimp Thanks for pointing out.

> > Some things that we have to consider prior to merging:
- [x] updated the commit message to add `BREAKING CHANGE` https://github.com/kentcdodds/eslint-find-new-rules/pull/27/commits/ea617cfc67cd222764207cc377865e9df49eb154
- [x] updated the usage documentation

Please have look, and let me know, if any changes required. Thanks!!

**Comment by kentcdodds** [comment]
LGTM

**Comment by ta2edchimp** [comment]
LGTM too.
Did a dry run on `semantic-release` and it works as I hoped.

:+1:
